packages/squiggle-lang/src/rescript/Reducer/Reducer_Expression/Reducer_Expression.res

```diff
@@ -38,50 +39,137 @@ let parse_ = (expr: string, parser, converter): result<t, errorValue> =>
+        switch expression {
+        | T.EList(list{T.EValue(EvCall("$let")), ..._}) => REExpressionExpected->Error
+        | _ => replaceSymbols(expression, bindings)
+        }
+      }
+
+      switch list {
+      | list{T.EValue(EvCall("$$bindings"))} => bindings->EBindings->Ok
+
+      | list{T.EValue(EvCall("$$bindStatement")), T.EBindings(bindings), statement} =>
+        doBindStatement(statement, bindings)
+      | list{T.EValue(EvCall("$$bindExpression")), T.EBindings(bindings), expression} =>
+        doBindExpression(expression, bindings)
+      | _ => list->T.EList->Ok
+      }
+    }
+
+    list->dispatchMacroCall(bindings)
   }
 
-/*
-  Recursively evaluate/reduce the code tree
-*/
-let rec reduceExpression = (expression: t, bindings): result<expressionValue, 'e> =>
-  switch expression {
-  | T.EValue(value) => value->Ok
-  | T.EList(list) => {
-      let racc: result<list<expressionValue>, 'e> = list->Belt.List.reduceReverse(Ok(list{}), (
-        racc,
-        each: expression,
-      ) =>
-        racc->Result.flatMap(acc => {
-          each
-          ->reduceExpression(bindings)
-          ->Result.flatMap(newNode => {
-            acc->Belt.List.add(newNode)->Ok
+  let rec seekMacros = (expression: t, bindings: T.bindings): result<t, 'e> =>
+    switch expression {
+    | T.EValue(value) => expression->Ok
+    | T.EList(list) => {
+        let racc: result<list<t>, 'e> = list->Belt.List.reduceReverse(Ok(list{}), (
+          racc,
+          each: expression,
+        ) =>
+          racc->Result.flatMap(acc => {
+            each
+            ->seekMacros(bindings)
+            ->Result.flatMap(newNode => {
+              acc->Belt.List.add(newNode)->Ok
+            })
           })
-        })
-      )
-      racc->Result.flatMap(acc => acc->reduceValueList)
+        )
+        racc->Result.flatMap(acc => acc->doMacroCall(bindings))
+      }
     }
-  }
+
+  let rec reduceExpandedExpression = (expression: t): result<expressionValue, 'e> =>
+    switch expression {
+    | T.EValue(value) => value->Ok
+    | T.EList(list) => {
+        let racc: result<list<expressionValue>, 'e> = list->Belt.List.reduceReverse(Ok(list{}), (
+          racc,
+          each: expression,
+        ) =>
+          racc->Result.flatMap(acc => {
+            each
+            ->reduceExpandedExpression
+            ->Result.flatMap(newNode => {
+              acc->Belt.List.add(newNode)->Ok
+            })
+          })
+        )
+        racc->Result.flatMap(acc => acc->reduceValueList)
+      }
+    }
+
+  let rExpandedExpression: result<t, 'e> = expression->seekMacros(bindings)
+  rExpandedExpression->Result.flatMap(expandedExpression =>
+    expandedExpression->reduceExpandedExpression
+  )
+}
 
 let evalWBindingsExpression = (aExpression, bindings): result<expressionValue, 'e> =>
   reduceExpression(aExpression, bindings)
 
 /*
   Evaluates MathJs code via Lisp using bindings and answers the result
```

Comment:
could you call the interpreter `reducer` here instead of `lisp`?

Reply:
umuro/squiggle#50